src/main/java/dev/openfeature/sdk/LayeredEvaluationContext.java

```diff
@@ -0,0 +1,235 @@
+        } else if (apiContext != null && apiContext.getTargetingKey() != null) {
+            this.targetingKey = apiContext.getTargetingKey();
+        } else {
+            this.targetingKey = null;
+        }
+    }
+
+    @Override
+    public String getTargetingKey() {
+        return targetingKey;
+    }
+
+    /**
+     * Using this method should be avoided as it comes with a performance cost.
+     * Consider constructing a new LayeredEvaluationContext instead.
+     *
+     * <p>
+     * Does not modify this object.
+     *
+     * @param overridingContext overriding context
+     * @return A new LayeredEvaluationContext containing the context from this object, with the overridingContext
+     *      merged on top.
+     * @deprecated Use of this method is discouraged due to performance considerations.
+     */
+    @Deprecated
+    @Override
+    public EvaluationContext merge(EvaluationContext overridingContext) {
+        var merged = new LayeredEvaluationContext(apiContext, transactionContext, clientContext, invocationContext);
+
+        if (this.hookContexts == null) {
+            merged.hookContexts = new ArrayList<>(1);
+        } else {
+            merged.hookContexts = new ArrayList<>(this.hookContexts.size() + 1);
+            merged.hookContexts.addAll(this.hookContexts);
+        }
+        merged.hookContexts.add(overridingContext);
+
+        var otherTargetingKey = overridingContext.getTargetingKey();
+        if (otherTargetingKey != null) {
+            merged.targetingKey = otherTargetingKey;
+        }
+        return merged;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return (hookContexts == null || hookContexts.isEmpty())
+                && (invocationContext == null || invocationContext.isEmpty())
+                && (clientContext == null || clientContext.isEmpty())
+                && (transactionContext == null || transactionContext.isEmpty())
+                && (apiContext == null || apiContext.isEmpty());
+    }
+
+    @Override
+    public Set<String> keySet() {
+        return Collections.unmodifiableSet(ensureKeySet());
+    }
+
+    private Set<String> ensureKeySet() {
+        if (this.keySet != null) {
+            return this.keySet;
+        }
+
+        var keys = new HashSet<String>();
+
+        if (hookContexts != null) {
+            for (int i = 0; i < hookContexts.size(); i++) {
+                var current = hookContexts.get(i);
+                keys.addAll(current.keySet());
+            }
+        }
+
+        if (invocationContext != null) {
+            keys.addAll(invocationContext.keySet());
+        }
+        if (clientContext != null) {
+            keys.addAll(clientContext.keySet());
+        }
+        if (transactionContext != null) {
+            keys.addAll(transactionContext.keySet());
+        }
+        if (apiContext != null) {
+            keys.addAll(apiContext.keySet());
+        }
+        this.keySet = keys;
```

Comment:
Should we make use of the `Collections.emptySet()` singleton while we have no keys?

Reply:
It's the allocation of the `HashSet` on 107 that I think we can save, possibly, which is a heap object... I think we can do all the null checks and then only create this object if any of them are non null, otherwise just return the singleton, right? Maybe I'm missing something.

Reply:
could  the `isEmpty()` method be used to determine whether to create the `HashSet`, or if false, return the singleton?